ruby/compress-image/index.rb

```diff
@@ -0,0 +1,107 @@
+require "json"
+require 'open-uri'
+require "base64"
+require "tinify"
+require "kraken-io"
+
+class NotImageException < StandardError; end
+class ApiError < StandardError; end
+
+invalidPayload = {
+  success: false, message: "Invalid Payload"
+}
+notImage = {
+  success: false, message: "Input file is not an image."
+}
+invalidApi = {
+  success: false, message: "Invalid API."
+}
+
+def tinypng(req, image) 
+    # Get api key
+    if !req.env['TINYPNG_API']
+      raise ApiError
+    end
+
+    Tinify.key = req.env['TINYPNG_API']
+    buffer = Base64.decode64(image)
+    result = Tinify.from_buffer(buffer).to_buffer
+    
+    Base64.encode64(result)
+end
+
+def krakenio(req, image) 
+    if !req.env['KRAKENIO_KEY'] or !req.env['KRAKENIO_SECRET']
+        raise ApiError
+    end
+
+    buffer = Base64.decode64(image)
+    File.open('/file.png', 'wb') do |f|
+      f.write(buffer)
+    end
+
+    kraken = Kraken::API.new(
+        :api_key => req.env['KRAKENIO_KEY'],
+        :api_secret => req.env['KRAKENIO_SECRET']
+    )
+
+    data = kraken.upload('/file.png')
+
+    if data.success
+        img = URI.open(data.kraked_url)
+        Base64.encode64(img.read)
+    else
+        raise NotImageException
+    end
+end
+
+def main(req, res)
+    # Input validation
+    provider = nil
+    image = nil
+    begin
+        payload = JSON.parse(req.payload)
+        provider = payload["provider"]
+        image = payload["image"]
+    rescue Exception => err
+        return res.json(invalidPayload)
+    end
+
+    if provider.nil? or provider.empty? or image.nil? or image.empty?
+        return res.json(invalidPayload)
```

Comment:
I tried to execute: 

```shell
curl -H "X-Internal-Challenge: secret-key" -H "Content-Type: application/json" -X POST http://localhost:3000/ -d '{"payload": "{\"image\": \"\", \"provider\":\"tinypng\"}"}'
```

but got this error:

```json
{
  "code": 500,
  "message": "/usr/code-start/index.rb:71:in `main'\n/usr/local/src/server.rb:89:in `block in <top (required)>'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1684:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1684:in `block in compile!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1002:in `block (3 levels) in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1020:in `route_eval'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1002:in `block (2 levels) in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1051:in `block in process_route'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1049:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1049:in `process_route'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1000:in `block in route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:997:in `each'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:997:in `route!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1121:in `block in dispatch!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `invoke'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1116:in `dispatch!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:932:in `block in call!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `catch'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1092:in `invoke'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:932:in `call!'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:921:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/xss_header.rb:20:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/path_traversal.rb:18:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/json_csrf.rb:28:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/base.rb:53:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/base.rb:53:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-protection-3.0.2/lib/rack/protection/frame_options.rb:33:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/logger.rb:17:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/common_logger.rb:38:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:261:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:254:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/head.rb:12:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/method_override.rb:24:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:219:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1995:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1554:in `block in call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1770:in `synchronize'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/sinatra-3.0.2/lib/sinatra/base.rb:1554:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/configuration.rb:268:in `call'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:89:in `block in handle_request'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:343:in `with_force_shutdown'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/request.rb:88:in `handle_request'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:430:in `process_client'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/server.rb:233:in `block in run'\n/usr/local/src/vendor/bundle/ruby/3.1.0/gems/puma-6.0.0/lib/puma/thread_pool.rb:150:in `block in spawn_thread'"
}
```

The runtime container logs show:

```
#<NameError: undefined local variable or method `invalidPayload' for #<Sinatra::Application:0x00007fdda97c0680 @default_layout=:layout, @preferred_extension=nil, @app=nil, @template_cache=#<Tilt::Cache:0x00007fddaad9c8c8 @cache={}>, @pinned_response=true, @env={"rack.version"=>[1, 6], "rack.errors"=>#<IO:<STDERR>>, "rack.multithread"=>true, "rack.multiprocess"=>false, "rack.run_once"=>false, "rack.url_scheme"=>"http", "SCRIPT_NAME"=>"", "QUERY_STRING"=>"", "SERVER_SOFTWARE"=>"puma 6.0.0 Sunflower", "GATEWAY_INTERFACE"=>"CGI/1.2", "REQUEST_METHOD"=>"POST", "REQUEST_PATH"=>"/", "REQUEST_URI"=>"/", "SERVER_PROTOCOL"=>"HTTP/1.1", "HTTP_HOST"=>"localhost:3000", "HTTP_USER_AGENT"=>"curl/7.68.0", "HTTP_ACCEPT"=>"*/*", "HTTP_X_INTERNAL_CHALLENGE"=>"secret-key", "CONTENT_TYPE"=>"application/json", "CONTENT_LENGTH"=>"58", "puma.request_body_wait"=>0.02228003740310669, "SERVER_NAME"=>"localhost", "SERVER_PORT"=>"3000", "PATH_INFO"=>"/", "REMOTE_ADDR"=>"172.17.0.1", "HTTP_VERSION"=>"HTTP/1.1", "puma.socket"=>#<TCPSocket:fd 11, AF_INET, 172.17.0.2, 3000>, "rack.hijack?"=>true, "rack.hijack"=>#<Puma::Client:0xc954 @ready=true>, "rack.input"=>#<StringIO:0x00007fdda97c36c8>, "rack.after_reply"=>[], "puma.config"=>#<Puma::Configuration:0x00007fddab518e58 @options=#<Puma::UserFileDefaultOptions:0x00007fddab5183b8 @user_options={:binds=>["tcp://0.0.0.0:3000"], :environment=>"production", :log_writer=>#<Puma::LogWriter:0x00007fddab519330 @formatter=#<Puma::LogWriter::DefaultFormatter:0x00007fddab5192e0>, @stdout=#<IO:<STDOUT>>, @stderr=#<IO:<STDERR>>, @debug=false, @error_logger=#<Puma::ErrorLogger:0x00007fddab519240 @ioerr=#<IO:<STDERR>>, @debug=false>>}, @file_options={}, @default_options={:auto_trim_time=>30, :binds=>["tcp://0.0.0.0:9292"], :clean_thread_locals=>false, :debug=>false, :early_hints=>nil, :environment=>"development", :first_data_timeout=>30, :io_selector_backend=>:auto, :log_requests=>false, :logger=>#<IO:<STDOUT>>, :max_fast_inline=>10, :max_threads=>5, :min_threads=>0, :mode=>:http, :mutate_stdout_and_stderr_to_sync_on_write=>true, :out_of_band=>[], :persistent_timeout=>20, :queue_requests=>true, :rackup=>"config.ru", :raise_exception_on_sigterm=>true, :reaping_time=>1, :remote_address=>:socket, :silence_single_worker_warning=>false, :tag=>"src", :tcp_host=>"0.0.0.0", :tcp_port=>9292, :wait_for_less_busy_worker=>0.005, :worker_boot_timeout=>60, :worker_check_interval=>5, :worker_culling_strategy=>:youngest, :worker_shutdown_timeout=>30, :worker_timeout=>60, :workers=>0, :preload_app=>false}>, @plugins=#<Puma::PluginLoader:0x00007fddab518278 @instances=[]>, @user_dsl=#<Puma::DSL:0x00007fddab518138 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={:binds=>["tcp://0.0.0.0:3000"], :environment=>"production", :log_writer=>#<Puma::LogWriter:0x00007fddab519330 @formatter=#<Puma::LogWriter::DefaultFormatter:0x00007fddab5192e0>, @stdout=#<IO:<STDOUT>>, @stderr=#<IO:<STDERR>>, @debug=false, @error_logger=#<Puma::ErrorLogger:0x00007fddab519240 @ioerr=#<IO:<STDERR>>, @debug=false>>}, @plugins=[]>, @file_dsl=#<Puma::DSL:0x00007fddab518048 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={}, @plugins=[]>, @default_dsl=#<Puma::DSL:0x00007fddab51fde8 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={:auto_trim_time=>30, :binds=>["tcp://0.0.0.0:9292"], :clean_thread_locals=>false, :debug=>false, :early_hints=>nil, :environment=>"development", :first_data_timeout=>30, :io_selector_backend=>:auto, :log_requests=>false, :logger=>#<IO:<STDOUT>>, :max_fast_inline=>10, :max_threads=>5, :min_threads=>0, :mode=>:http, :mutate_stdout_and_stderr_to_sync_on_write=>true, :out_of_band=>[], :persistent_timeout=>20, :queue_requests=>true, :rackup=>"config.ru", :raise_exception_on_sigterm=>true, :reaping_time=>1, :remote_address=>:socket, :silence_single_worker_warning=>false, :tag=>"src", :tcp_host=>"0.0.0.0", :tcp_port=>9292, :wait_for_less_busy_worker=>0.005, :worker_boot_timeout=>60, :worker_check_interval=>5, :worker_culling_strategy=>:youngest, :worker_shutdown_timeout=>30, :worker_timeout=>60, :workers=>0, :preload_app=>false}, @plugins=[]>>, "sinatra.commonlogger"=>true, "rack.logger"=>#<Logger:0x00007fdda97c1148 @level=1, @progname=nil, @default_formatter=#<Logger::Formatter:0x00007fdda97c1008 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x00007fdda97c0c70 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDERR>>, @binmode=false, @mon_data=#<Monitor:0x00007fdda97c0ae0>, @mon_data_owner_object_id=51520>>, "rack.request.query_string"=>"", "rack.request.query_hash"=>{}, "sinatra.route"=>"POST /"}, @params={}, @request=#<Sinatra::Request:0x00007fdda97c0338 @params={}, @env={"rack.version"=>[1, 6], "rack.errors"=>#<IO:<STDERR>>, "rack.multithread"=>true, "rack.multiprocess"=>false, "rack.run_once"=>false, "rack.url_scheme"=>"http", "SCRIPT_NAME"=>"", "QUERY_STRING"=>"", "SERVER_SOFTWARE"=>"puma 6.0.0 Sunflower", "GATEWAY_INTERFACE"=>"CGI/1.2", "REQUEST_METHOD"=>"POST", "REQUEST_PATH"=>"/", "REQUEST_URI"=>"/", "SERVER_PROTOCOL"=>"HTTP/1.1", "HTTP_HOST"=>"localhost:3000", "HTTP_USER_AGENT"=>"curl/7.68.0", "HTTP_ACCEPT"=>"*/*", "HTTP_X_INTERNAL_CHALLENGE"=>"secret-key", "CONTENT_TYPE"=>"application/json", "CONTENT_LENGTH"=>"58", "puma.request_body_wait"=>0.02228003740310669, "SERVER_NAME"=>"localhost", "SERVER_PORT"=>"3000", "PATH_INFO"=>"/", "REMOTE_ADDR"=>"172.17.0.1", "HTTP_VERSION"=>"HTTP/1.1", "puma.socket"=>#<TCPSocket:fd 11, AF_INET, 172.17.0.2, 3000>, "rack.hijack?"=>true, "rack.hijack"=>#<Puma::Client:0xc954 @ready=true>, "rack.input"=>#<StringIO:0x00007fdda97c36c8>, "rack.after_reply"=>[], "puma.config"=>#<Puma::Configuration:0x00007fddab518e58 @options=#<Puma::UserFileDefaultOptions:0x00007fddab5183b8 @user_options={:binds=>["tcp://0.0.0.0:3000"], :environment=>"production", :log_writer=>#<Puma::LogWriter:0x00007fddab519330 @formatter=#<Puma::LogWriter::DefaultFormatter:0x00007fddab5192e0>, @stdout=#<IO:<STDOUT>>, @stderr=#<IO:<STDERR>>, @debug=false, @error_logger=#<Puma::ErrorLogger:0x00007fddab519240 @ioerr=#<IO:<STDERR>>, @debug=false>>}, @file_options={}, @default_options={:auto_trim_time=>30, :binds=>["tcp://0.0.0.0:9292"], :clean_thread_locals=>false, :debug=>false, :early_hints=>nil, :environment=>"development", :first_data_timeout=>30, :io_selector_backend=>:auto, :log_requests=>false, :logger=>#<IO:<STDOUT>>, :max_fast_inline=>10, :max_threads=>5, :min_threads=>0, :mode=>:http, :mutate_stdout_and_stderr_to_sync_on_write=>true, :out_of_band=>[], :persistent_timeout=>20, :queue_requests=>true, :rackup=>"config.ru", :raise_exception_on_sigterm=>true, :reaping_time=>1, :remote_address=>:socket, :silence_single_worker_warning=>false, :tag=>"src", :tcp_host=>"0.0.0.0", :tcp_port=>9292, :wait_for_less_busy_worker=>0.005, :worker_boot_timeout=>60, :worker_check_interval=>5, :worker_culling_strategy=>:youngest, :worker_shutdown_timeout=>30, :worker_timeout=>60, :workers=>0, :preload_app=>false}>, @plugins=#<Puma::PluginLoader:0x00007fddab518278 @instances=[]>, @user_dsl=#<Puma::DSL:0x00007fddab518138 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={:binds=>["tcp://0.0.0.0:3000"], :environment=>"production", :log_writer=>#<Puma::LogWriter:0x00007fddab519330 @formatter=#<Puma::LogWriter::DefaultFormatter:0x00007fddab5192e0>, @stdout=#<IO:<STDOUT>>, @stderr=#<IO:<STDERR>>, @debug=false, @error_logger=#<Puma::ErrorLogger:0x00007fddab519240 @ioerr=#<IO:<STDERR>>, @debug=false>>}, @plugins=[]>, @file_dsl=#<Puma::DSL:0x00007fddab518048 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={}, @plugins=[]>, @default_dsl=#<Puma::DSL:0x00007fddab51fde8 @config=#<Puma::Configuration:0x00007fddab518e58 ...>, @options={:auto_trim_time=>30, :binds=>["tcp://0.0.0.0:9292"], :clean_thread_locals=>false, :debug=>false, :early_hints=>nil, :environment=>"development", :first_data_timeout=>30, :io_selector_backend=>:auto, :log_requests=>false, :logger=>#<IO:<STDOUT>>, :max_fast_inline=>10, :max_threads=>5, :min_threads=>0, :mode=>:http, :mutate_stdout_and_stderr_to_sync_on_write=>true, :out_of_band=>[], :persistent_timeout=>20, :queue_requests=>true, :rackup=>"config.ru", :raise_exception_on_sigterm=>true, :reaping_time=>1, :remote_address=>:socket, :silence_single_worker_warning=>false, :tag=>"src", :tcp_host=>"0.0.0.0", :tcp_port=>9292, :wait_for_less_busy_worker=>0.005, :worker_boot_timeout=>60, :worker_check_interval=>5, :worker_culling_strategy=>:youngest, :worker_shutdown_timeout=>30, :worker_timeout=>60, :workers=>0, :preload_app=>false}, @plugins=[]>>, "sinatra.commonlogger"=>true, "rack.logger"=>#<Logger:0x00007fdda97c1148 @level=1, @progname=nil, @default_formatter=#<Logger::Formatter:0x00007fdda97c1008 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x00007fdda97c0c70 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<IO:<STDERR>>, @binmode=false, @mon_data=#<Monitor:0x00007fdda97c0ae0>, @mon_data_owner_object_id=51520>>, "rack.request.query_string"=>"", "rack.request.query_hash"=>{}, "sinatra.route"=>"POST /"}>, @response=#<Sinatra::Response:0x00007fdda97c0248 @status=200, @headers={"Content-Type"=>"application/json"}, @writer=#<Method: Sinatra::Response(Rack::Response::Helpers)#append(chunk) /usr/local/src/vendor/bundle/ruby/3.1.0/gems/rack-2.2.4/lib/rack/response.rb:287>, @block=nil, @body=[], @buffered=true, @length=0>>

        return res.json(invalidPayload)
                        ^^^^^^^^^^^^^^>
```

This seems to happen for `invalidPayload`, `notImage`, and `invalidApi`.

Reply:
I don't get it. What would be the expected behavior?

Reply:
It should return something like {success: false, message: <descriptive message>} in the response.